xarray/core/common.py

```diff
@@ -718,11 +719,27 @@ def assign_attrs(self, *args: Any, **kwargs: Any) -> Self:
         out.attrs.update(*args, **kwargs)
         return out
 
+    @overload
+    def pipe(
+        self,
+        func: Callable[Concatenate[Self, P], T],
+        *args: P.args,
+        **kwargs: P.kwargs,
+    ) -> T: ...
+
+    @overload
     def pipe(
         self,
-        func: Callable[..., T] | tuple[Callable[..., T], str],
+        func: tuple[Callable[..., T], str],
```

Comment:
Why not this as well?
```suggestion
        func: tuple[Callable[P, T], str]
```

Reply:
We cannot do that because when we pass the function as part of a tuple we don't have enough information to precisely type the function's parameters.

In your suggested change, the ParamSpec `P` represents _all_ of the function's parameters, but we need to know all of the parameters _excluding_ one (the one that takes the data value, as identified by the name given in the second value of the tuple).

If that's as clear as mud, let me try to clarify with more detail.

In the first form, where we pass only a function as the first argument to `pipe`, we expect the function to take the data value as the _first_ argument, meaning that we know exactly where in the function's list of parameters the data parameter is: the first position.

This means we can type the function more precisely, like so, indicating that the data parameter is first, concatenated with zero or more positional and keyword parameters (and returning a value of some type `T`):

```
# Self is a DataWithCoords (DataArray, Dataset, others?), or DataTree
# P represents all parameters to func *excluding* Self
func: Callable[Concatenate[Self, P], T]
```

Therefore, after passing `func` to `pipe`, we must pass all arguments _except_ for the data (self) argument, and this is represented by `P`, which _excludes_ `Self`:

```
# pipe expects a function followed by all arguments to pass to the function,
# *except* for the data argument, which pipe will *implicitly* pass.
def pipe(f: Callable[Concatenate[Self, P], T], *args: P.args, **kwargs: P.kwargs) -> T: ...
```

However, when we pass a function/keyword 2-tuple as the first argument to `pipe`, we have no idea what position the keyword parameter is in `func`'s type signature. We only know that it's _not_ the first parameter.

This means, we cannot do the following, as you suggest, because in this case `P` _includes_ the data parameter, but it must not, and there's no way of omitting it without knowing precisely what position it's in:

```
# We don't know where Self is within P, so we have no way of defining P
# as meaning all of func's parameters *except* for Self. Therefore, this
# signature indicates that we can *explicity* pass a data argument, but that's
# not correct.
def pipe(func: tuple[Callable[P, T], str], *args: P.args, **kwargs: P.kwargs) -> T: ...
```

To clarify, although `Callable[P, T]` is valid for the function itself within the tuple, using `*args: P.args, **kwargs: P.kwargs` for the rest of the parameters to `pipe` in this case is incorrect, because it means that we can _explicitly_ pass the data argument in the mix there (because `P` _includes_ `Self`), but the whole point of `pipe`, of course, is to _implicitly_ pass the data argument, and thus _not_ allow it to be passed explicitly.

Technically speaking, as far as mypy is concerned, your suggestion probably make no difference from what I propose, but in terms of the information it conveys to the reader, it is incorrect.

This is why I discourage the use of the tuple form, and instead recommend the use of a `lambda` (or another function def with args reordered such that the data arg is first). Even using a `lambda` to reorder things allows mypy to be more helpful than is possible with the tuple form.

For example (taken from the test cases in `xarray/tests/test_dataset_typing.yml` in this PR):

```
    from xarray import Dataset

    def f(arg: int, ds: Dataset) -> Dataset:
        return ds

    # Since we cannot provide a precise type annotation when passing a tuple to
    # pipe, there's not enough information for type analysis to indicate that
    # we are missing an argument for parameter `arg`, so we get no error here.

    ds = Dataset().pipe((f, "ds"))
    reveal_type(ds)  # N: Revealed type is "xarray.core.dataset.Dataset"

    # Rather than passing a tuple, passing a lambda that calls `f` with args in
    # the correct order allows for proper type analysis, indicating (perhaps
    # somewhat cryptically) that we failed to pass an argument for `arg`.

    ds = Dataset().pipe(lambda data, arg: f(arg, data))

    # mypy produces the following output for the line above, as it should,
    # indicating that we forgot to pass an argument to pipe, which pipe needs
    # to pass to `f` in the 2nd position.
    error: No overload variant of "pipe" of "DataWithCoords" matches argument type "Callable[[Any, Any], Dataset]"  [call-overload]
    note: Possible overload variants:
    note:     def [P`9, T] pipe(self, func: Callable[[Dataset, **P], T], *args: P.args, **kwargs: P.kwargs) -> T
    note:     def [T] pipe(self, func: tuple[Callable[..., T], str], *args: Any, **kwargs: Any) -> T
 ```

IMHO, the tuple form should not be supported, for this very reason, but I don't expect that deprecating that form would get much traction from anybody else.